File: src/test/java/com/zepben/cimbend/network/tracing/FindWithUsagePointsTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 * This file is part of evolve-sdk-jvm.
 *
 * evolve-sdk-jvm is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * evolve-sdk-jvm is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with evolve-sdk-jvm.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.zepben.cimbend.network.tracing;

import com.zepben.cimbend.cim.iec61968.metering.UsagePoint;
import com.zepben.cimbend.cim.iec61970.base.core.ConductingEquipment;
import com.zepben.cimbend.cim.iec61970.base.wires.PowerTransformer;
import com.zepben.cimbend.network.NetworkService;
import com.zepben.cimbend.testdata.TestNetworks;
import com.zepben.test.util.junit.SystemLogExtension;
import kotlin.jvm.JvmField;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static com.zepben.cimbend.network.tracing.FindWithUsagePoints.Result.Status.*;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

public class FindWithUsagePointsTest {

    @JvmField
    @RegisterExtension
    public SystemLogExtension systemOutRule = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess();

    private final FindWithUsagePoints findWithUsagePoints = new FindWithUsagePoints();

    //
    //      c0   c1   c2             c3   c4   c5
    //  es ----+----+----+- sw1 iso ----+----+---- tx1
    //         |    |    |              |    |
    //         |c6  |c7  |c8            |c9  |c10
    //         |    |    |              |    |
    //        tx2  tx3   |             tx4  tx5
    //                   |
    //                   |       c11  c12  c13
    //                   +- sw2 ----+----+---- tx6
    //                              |    |
    //                              |c14 |c15
    //                              |    |
    //                             tx7  tx8
    //
    // sw1: normally close, currently open
    // sw2: normally open, currently closed
    //
    private final NetworkService network = TestNetworks.withUsagePointsNetwork();

    private final ConductingEquipment es = ce("es");
    private final ConductingEquipment c1 = ce("c1");
    private final ConductingEquipment c3 = ce("c3");
    private final ConductingEquipment c5 = ce("c5");
    private final ConductingEquipment c11 = ce("c11");
    private final ConductingEquipment c12 = ce("c12");
    private final ConductingEquipment tx6 = ce("tx6");

    @Test
    public void normalStateSingleTrace() {
        validate(findWithUsagePoints.runNormal(es, null), NO_ERROR, "tx1", "tx2", "tx3", "tx4", "tx5", "iso");
        validate(findWithUsagePoints.runNormal(c3, c5), NO_ERROR, "tx4", "tx5");
        validate(findWithUsagePoints.runNormal(c5, c3), NO_ERROR, "tx4", "tx5");
        validate(findWithUsagePoints.runNormal(es, c11), NO_PATH);
    }

    @Test
    public void normalStateMultiTrace() {
        List<FindWithUsagePoints.Result> results = findWithUsagePoints.runNormal(Arrays.asList(es, c3, c5, es), Arrays.asList(null, c5, c3, c11));

        assertThat(results.size(), equalTo(4));
        validate(results.get(0), NO_ERROR, "tx1", "tx2", "tx3", "tx4", "tx5", "iso");
        validate(results.get(1), NO_ERROR, "tx4", "tx5");
        validate(results.get(2), NO_ERROR, "tx4", "tx5");
        validate(results.get(3), NO_PATH);

        validateMismatch(findWithUsagePoints.runNormal(Arrays.asList(es, es), Collections.singletonList(null)), 2);
        validateMismatch(findWithUsagePoints.runNormal(Arrays.asList(es, es), Arrays.asList(c1, c11, c3)), 3);
    }

    @Test
    public void currentStateSingleTrace() {
        validate(findWithUsagePoints.runCurrent(es, null), NO_ERROR, "tx2", "tx3", "tx6", "tx7", "tx8");
        validate(findWithUsagePoints.runCurrent(c1, c12), NO_ERROR, "tx3", "tx7");
        validate(findWithUsagePoints.runCurrent(c12, c1), NO_ERROR, "tx3", "tx7");
        validate(findWithUsagePoints.runCurrent(es, c5), NO_PATH);
    }

    @Test
    public void currentStateMultiTrace() {
        List<FindWithUsagePoints.Result> results = findWithUsagePoints.runCurrent(Arrays.asList(es, c1, c12, es), Arrays.asList(null, c12, c1, c5));

        assertThat(results.size(), equalTo(4));
        validate(results.get(0), NO_ERROR, "tx2", "tx3", "tx6", "tx7", "tx8");
        validate(results.get(1), NO_ERROR, "tx3", "tx7");
        validate(results.get(2), NO_ERROR, "tx3", "tx7");
        validate(results.get(3), NO_PATH);

        validateMismatch(findWithUsagePoints.runCurrent(Arrays.asList(es, es), Collections.singletonList(null)), 2);
        validateMismatch(findWithUsagePoints.runCurrent(Arrays.asList(es, es), Arrays.asList(c1, c11, c3)), 3);
    }

    @Test
    public void sameFromAndTo() {
        validate(findWithUsagePoints.runNormal(c3, c3), NO_ERROR);
        validate(findWithUsagePoints.runNormal(tx6, tx6), NO_ERROR, "tx6");
    }

    @Test
    public void worksWithNoTerminals() {
        PowerTransformer tx1 = new PowerTransformer();
        PowerTransformer tx2 = new PowerTransformer();
        UsagePoint usagePoint = new UsagePoint();

        usagePoint.addEquipment(tx1);
        tx1.addUsagePoint(usagePoint);

        validate(findWithUsagePoints.runNormal(tx1, null), NO_ERROR, tx1.getMRID());
        validate(findWithUsagePoints.runNormal(tx1, tx1), NO_ERROR, tx1.getMRID());
        validate(findWithUsagePoints.runNormal(tx2, null), NO_ERROR);
        validate(findWithUsagePoints.runNormal(tx2, tx2), NO_ERROR);
    }

    private ConductingEquipment ce(String mRID) {
        return network.get(ConductingEquipment.class, mRID);
    }

    private void validate(FindWithUsagePoints.Result result, FindWithUsagePoints.Result.Status expectedStatus, String... expectedMRIDs) {
        assertThat(result.status(), equalTo(expectedStatus));
        assertThat(result.conductingEquipment().keySet(), containsInAnyOrder(expectedMRIDs));
    }

    private void validateMismatch(List<FindWithUsagePoints.Result> results, int expectedResults) {
        assertThat(results.size(), equalTo(expectedResults));

        results.forEach(result -> {
            assertThat(result.status(), equalTo(MISMATCHED_FROM_TO));
            assertThat(result.conductingEquipment().keySet(), empty());
        });
    }

}

File: src/test/kotlin/com/zepben/evolve/services/network/tracing/FindWithUsagePointsTest.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.evolve.services.network.tracing

import com.zepben.evolve.cim.iec61968.metering.UsagePoint
import com.zepben.evolve.cim.iec61970.base.core.ConductingEquipment
import com.zepben.evolve.cim.iec61970.base.core.Terminal
import com.zepben.evolve.cim.iec61970.base.equivalents.EquivalentBranch
import com.zepben.evolve.cim.iec61970.base.wires.AcLineSegment
import com.zepben.evolve.cim.iec61970.base.wires.EnergySource
import com.zepben.evolve.cim.iec61970.base.wires.PowerTransformer
import com.zepben.evolve.services.network.testdata.SingleTransformerNetwork
import com.zepben.evolve.services.network.testdata.WithUsagePointsNetwork
import com.zepben.evolve.services.network.tracing.FindWithUsagePoints.Result.Status.NO_ERROR
import com.zepben.evolve.services.network.tracing.FindWithUsagePoints.Result.Status.NO_PATH
import com.zepben.evolve.services.network.tracing.FindWithUsagePoints.VirtualUsagePointCondition.*
import com.zepben.evolve.services.network.tracing.networktrace.operators.NetworkStateOperators
import com.zepben.testutils.junit.SystemLogExtension
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension


internal class FindWithUsagePointsTest {

    @JvmField
    @RegisterExtension
    var systemErr: SystemLogExtension = SystemLogExtension.SYSTEM_ERR.captureLog().muteOnSuccess()

    private val findWithUsagePoints = FindWithUsagePoints(NetworkStateOperators.NORMAL)

    //
    //      c0   c1   c2             c3   c4   c5
    //  es ----+----+----+- sw1 iso ----+----+---- tx1
    //         |    |    |              |    |
    //         |c6  |c7  |c8            |c9  |c10
    //         |    |    |              |    |
    //        tx2  tx3   |             tx4  tx5
    //                   |
    //                   |       c11  c12  c13
    //                   +- sw2 ----+----+---- tx6
    //                              |    |
    //                              |c14 |c15
    //                              |    |
    //                             tx7  tx8
    //
    // sw1: normally closed, currently open
    // sw2: normally open, currently closed
    //
    private val network = WithUsagePointsNetwork.createLarge()

    private val es = ce("es")
    private val c1 = ce("c1")
    private val c3 = ce("c3")
    private val c5 = ce("c5")
    private val c11 = ce("c11")
    private val c12 = ce("c12")
    private val tx6 = ce("tx6")

    @Test
    internal fun singleTrace() {
        validate(findWithUsagePoints.run(es, null), NO_ERROR, listOf("tx1", "tx2", "tx3", "tx4", "tx5", "iso"))
        validate(findWithUsagePoints.run(c3, c5), NO_ERROR, listOf("tx4", "tx5"))
        validate(findWithUsagePoints.run(c5, c3), NO_ERROR, listOf("tx4", "tx5"))
        validate(findWithUsagePoints.run(es, c11), NO_PATH, emptyList())
    }

    @Test
    internal fun multiTrace() {
        val results = findWithUsagePoints.run(listOf(es to null, c3 to c5, c5 to c3, es to c11))

        assertThat(results, hasSize(4))
        validate(results[0], NO_ERROR, listOf("tx1", "tx2", "tx3", "tx4", "tx5", "iso"))
        validate(results[1], NO_ERROR, listOf("tx4", "tx5"))
        validate(results[2], NO_ERROR, listOf("tx4", "tx5"))
        validate(results[3], NO_PATH, emptyList())
    }

    @Test
    internal fun sameFromAndTo() {
        validate(findWithUsagePoints.run(c3, c3), NO_ERROR, emptyList())
        validate(findWithUsagePoints.run(tx6, tx6), NO_ERROR, listOf("tx6"))
    }

    @Test
    internal fun worksWithNoTerminals() {
        val tx1 = PowerTransformer()
        val tx2 = PowerTransformer()
        val usagePoint = UsagePoint()

        usagePoint.addEquipment(tx1)
        tx1.addUsagePoint(usagePoint)

        validate(findWithUsagePoints.run(tx1, null), NO_ERROR, listOf(tx1.mRID))
        validate(findWithUsagePoints.run(tx1, tx1), NO_ERROR, listOf(tx1.mRID))
        validate(findWithUsagePoints.run(tx2, null), NO_ERROR, emptyList())
        validate(findWithUsagePoints.run(tx2, tx2), NO_ERROR, emptyList())
    }

    @Test
    internal fun doesntRelyOnTerminalSequenceNumbers() {
        val validateSingleTransformerNetwork = { sequenceNumber: Int ->
            val networkService1 = SingleTransformerNetwork.create(sequenceNumber)

            val s: EnergySource = networkService1["s"]!!
            val tx: PowerTransformer = networkService1["tx"]!!

            // Validate both `from` and `to` terminals.
            validate(findWithUsagePoints.run(tx, null), NO_ERROR, listOf(tx.mRID))
            validate(findWithUsagePoints.run(tx, s), NO_ERROR, listOf(tx.mRID))
        }

        validateSingleTransformerNetwork(1)
        validateSingleTransformerNetwork(2)
    }

    @Test
    internal fun noPathWithMissingTerminals() {
        validate(findWithUsagePoints.run(AcLineSegment(), AcLineSegment().apply { addTerminal(Terminal()) }), NO_PATH, emptyList())
        validate(findWithUsagePoints.run(AcLineSegment().apply { addTerminal(Terminal()) }, AcLineSegment()), NO_PATH, emptyList())
    }

    @Test
    internal fun ignoresLvWhenAggregated() {
        validateVirtualUsagePoints(LV_AGGREGATION_ONLY, listOf("tx", "ec2"), listOf("tx"), listOf("ec1", "ec2"))
        validateVirtualUsagePoints(NO_LV_AGGREGATION, listOf("ec1", "ec2"), listOf("ec1", "ec2"), listOf("tx", "ec1", "ec2"))
        validateVirtualUsagePoints(ALL, listOf("tx", "ec2"), listOf("tx"), listOf("tx", "ec1", "ec2"))
        validateVirtualUsagePoints(NONE, listOf("ec1", "ec2"), listOf("ec1", "ec2"), listOf("ec1", "ec2"))
    }

    @Test
    internal fun includesTransformerAndDownstream() {
        val ns = WithUsagePointsNetwork.createTxWithRealAndLv()
        validate(findWithUsagePoints.run(ns["es"]!!, null), NO_ERROR, listOf("tx", "ec"))
    }

    @Test
    internal fun lvAggregationExcludeLvNetwork() {
        val ns = WithUsagePointsNetwork.createTxWithRealAndLv("LV_AGGREGATION")
        validate(findWithUsagePoints.run(ns["es"]!!, null), NO_ERROR, listOf("tx"))
    }

    private fun ce(mRID: String): ConductingEquipment {
        return network.get(ConductingEquipment::class.java, mRID)!!
    }

    private fun validate(result: FindWithUsagePoints.Result, expectedStatus: FindWithUsagePoints.Result.Status, expectedMRIDs: List<String>) {
        assertThat(result.status, equalTo(expectedStatus))
        assertThat(result.conductingEquipment.keys, containsInAnyOrder(*expectedMRIDs.toTypedArray()))
    }

    private fun validateVirtualUsagePoints(
        virtualUsagePointCondition: FindWithUsagePoints.VirtualUsagePointCondition,
        expectedAclsMrids: List<String>,
        expectedBranchMrids: List<String>,
        expectedOtherMrids: List<String>
    ) {
        validateVirtualUsagePoints<AcLineSegment>(virtualUsagePointCondition, "LV_AGGREGATION", expectedAclsMrids)
        validateVirtualUsagePoints<EquivalentBranch>(virtualUsagePointCondition, "LV_AGGREGATION", expectedBranchMrids)
        validateVirtualUsagePoints<EquivalentBranch>(virtualUsagePointCondition, "OTHER", expectedOtherMrids)
    }

    private inline fun <reified T : ConductingEquipment> validateVirtualUsagePoints(
        virtualUsagePointCondition: FindWithUsagePoints.VirtualUsagePointCondition,
        virtualConnectionCategory: String,
        expectedMrids: List<String>
    ) {
        val ns = WithUsagePointsNetwork.createTxWithVirtual<T>(virtualConnectionCategory)
        validate(FindWithUsagePoints(NetworkStateOperators.NORMAL, virtualUsagePointCondition).run(ns["es"]!!, null), NO_ERROR, expectedMrids)
    }

}
